docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,105 @@
+* Find LDAP group members
+* Determine OpenShift `Identities` from LDAP group members
+* Determine OpenShift `Users` from OpenShift `Identities`
+* Determine OpenShift `Group` from LDAP group
+* Populate new OpenShift `Group`'s `Users` from resulting OpenShift `UserIdentityMappings`, update OpenShift `Group` metadata fields tied to LDAP attributes, leave other OpenShift `Group` fields unchanged
+
+##  Determining Which LDAP Groups To Sync
+An `LDAPGroupLister` determines what LDAP groups needs to be synced and outputs the result as a set of unique identifier strings. The other objects that take these unique identifiers need to understand the format of this string (e.g. these objects will be tightly coupled). For example, an `LDAPGroupLister` for schema 1 above cannot be used with a `LDAPGroupDataExtractor` for schema 2. The four `LDAPGroupLister` implementations that are necessary are:
+
+1. List all OpenShift `Groups` conforming to some label selector, minus a blacklist
+* List some whitelist of OpenShift `Groups`
+* List all LDAP groups, minus a blacklist
+* List some whitelist of LDAP groups
+
+```go
+// LDAPGroupLister lists the LDAP groups that need to be synced by a job. The LDAPGroupLister needs to
+// be paired with an LDAPGroupDataExtractor that understands the format of the unique identifiers returned
+// to represent the LDAP groups to be synced.
+type LDAPGroupLister interface {
+	ListGroups() (groupUIDs []string, err error)
+}
+```
+
+## Collecting LDAP Members and Metadata
+An `LDAPGroupDataExtractor` gathers information on an LDAP group based on a `groupUID`. It may cache LDAP responses for responsivity. The approach to implementing this structure will vary with LDAP schema as well as sync job request. There will be a specific implementation for each schema as well as another flavor of that implementation for that schema where all the LDAP groups need to be synced. Examples can be seen in [dataextractor.go](https://github.com/stevekuznetsov/origin/blob/skuznets/ldap-group-sync-proposal/pkg/cmd/experimental/sync-groups/dataextractor.go)
+
+```go
+// LDAPGroupDataExtractor is able to retrieve the list of members, labels, annotations and OpenShift Group
+// name for an LDAP group specified by a groupUID
+type LDAPGroupDataExtractor interface {
+  // ExtractMembers returns the list of LDAP first-class user entries that are members of the LDAP group
+  // specified by the groupUID
+	ExtractMembers(groupUID string) (members []*ldap.Entry, err error)
+  // ExtractLabels returns the set of labels that the OpenShift representation of the LDAP group specified
+  // by the groupUID should contain
+  ExtractLabels(groupUID string) (labels labels.Set, err error)
+  // ExtractAnnotations returns the mapping of annotation keys to values that the OpenShift group Representing
+  // the LDAP group specified by the groupUID should have
+  ExtractAnnotations(groupUID string) (annotations map[string]string, err error)
```

Comment:
> Also, this could live alone and would be self-sufficient in Schema 2.

Not sold.  This fits and flows pretty nicely here.  I think it's where the abstraction belongs.